<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Exceptions - C++ Tutorials</title>
<link rel="canonical" href="http://www.cplusplus.com/doc/tutorial/exceptions/">
<link rel="shortcut icon" type="image/x-icon" href="/favicon.ico">
<link rel="stylesheet" type="text/css" href="/v321/main.css">
<script src="/v321/main.js" type="text/javascript"></script>
<script type='text/javascript'>
var googletag = googletag || {};
googletag.cmd = googletag.cmd || [];
(function() {
var gads = document.createElement('script');
gads.async = true;
gads.type = 'text/javascript';
var useSSL = 'https:' == document.location.protocol;
gads.src = (useSSL ? 'https:' : 'http:') + 
'//www.googletagservices.com/tag/js/gpt.js';
var node = document.getElementsByTagName('script')[0];
node.parentNode.insertBefore(gads, node);
})();
</script>

<script type='text/javascript'>
googletag.cmd.push(function() {
googletag.defineSlot('/32882001/L', [728, 90], 'div-gpt-ad-1427191279638-0').addService(googletag.pubads());
googletag.enableServices();
});
</script>
</head>
<body>
<div id="I_top">
<div id="I_header">
<div id="I_logo"><a href="/" title="cplusplus.com"><div></div></a></div>
<div id="I_search">
<form id="search" action="/search.do" method="get">
Search: <input name="q" size="20" class="txt"> <input type="submit" value="Go" class="btn">
</form>
</div>
<div id="I_bar">
<ul>
<li><a href="/doc/">Tutorials</a></li>
<li><a href="/doc/tutorial/">C++ Language</a></li>
<li class="here">Exceptions</li>
</ul>
</div>
<div id="I_user" class="C_LoginBox"><span title="ajax"></span></div>
</div>
</div>
<div id="I_mid">
<div id="I_wrap">
<div id="I_minheight"></div>
<div id="I_main">
<div class="C_support">
<div id='div-gpt-ad-1427191279638-0' style='width:728px; height:90px;'>
<script type='text/javascript'>
googletag.cmd.push(function() { googletag.display('div-gpt-ad-1427191279638-0'); });
</script>
</div>
</div>
<div id="I_content">
<div class="C_doc">
<h1>Exceptions</h1>
<section id="exceptions">
Exceptions provide a way to react to exceptional circumstances (like runtime errors) in programs by transferring control to special functions called <i>handlers</i>.<br>
<br>
To catch exceptions, a portion of code is placed under exception inspection. This is done by enclosing that portion of code in a <i>try-block</i>. When an exceptional circumstance arises within that block, an exception is thrown that transfers the control to the exception handler. If no exception is thrown, the code continues normally and all handlers are ignored.<br>
<br>
An exception is thrown by using the <code>throw</code> keyword from inside the <code>try</code> block. Exception handlers are declared with the keyword <code>catch</code>, which must be placed immediately after the <code>try</code> block:<br>
<br>
<div class="auto"><table class="split"><tr><td class="rownum"><pre><code>1<br>2<br>3<br>4<br>5<br>6<br>7<br>8<br>9<br>10<br>11<br>12<br>13<br>14<br>15<br></code></pre></td>
<td class="source"><pre><code><cite>// exceptions</cite>
<dfn>#include &lt;iostream&gt;</dfn>
<var>using</var> <var>namespace</var> std;

<var>int</var> main () {
  <var>try</var>
  {
    <var>throw</var> 20;
  }
  <var>catch</var> (<var>int</var> e)
  {
    cout &lt;&lt; <kbd>"An exception occurred. Exception Nr. "</kbd> &lt;&lt; e &lt;&lt; <kbd>'\n'</kbd>;
  }
  <var>return</var> 0;
}</code></pre></td>
<td class="output"><pre><samp>An exception occurred. Exception Nr. 20</samp></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
<br>
The code under exception handling is enclosed in a <code>try</code> block. In this example this code simply throws an exception:<br>
<br>
<div class="auto"><table class="snippet"><tr><td class="rownum"><pre><code>&nbsp;</code></pre></td>
<td class="source"><pre><code><var>throw</var> 20;</code></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
<br>
A <code>throw</code> expression accepts one parameter (in this case the integer value <code>20</code>), which is passed as an argument to the exception handler.<br>
<br>
The exception handler is declared with the <code>catch</code> keyword immediately after the closing brace of the <code>try</code> block. The syntax for <code>catch</code> is similar to a regular function with one parameter. The type of this parameter is very important, since the type of the argument passed by the <code>throw</code> expression is checked against it, and only in the case they match, the exception is caught by that handler.<br>
<br>
Multiple handlers (i.e., <code>catch</code> expressions) can be chained; each one with a different parameter type. Only the handler whose argument type matches the type of the exception specified in the <code>throw</code> statement is executed.<br>
<br>
If an ellipsis (<code>...</code>) is used as the parameter of <code>catch</code>, that handler will catch any exception no matter what the type of the exception thrown. This can be used as a default handler that catches all exceptions not caught by other handlers:<br>
<br>
<div class="auto"><table class="snippet"><tr><td class="rownum"><pre><code>1<br>2<br>3<br>4<br>5<br>6<br></code></pre></td>
<td class="source"><pre><code><var>try</var> {
  <cite>// code here</cite>
}
<var>catch</var> (<var>int</var> param) { cout &lt;&lt; <kbd>"int exception"</kbd>; }
<var>catch</var> (<var>char</var> param) { cout &lt;&lt; <kbd>"char exception"</kbd>; }
<var>catch</var> (...) { cout &lt;&lt; <kbd>"default exception"</kbd>; }</code></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
<br>
In this case, the last handler would catch any exception thrown of a type that is neither <code>int</code> nor <code>char</code>.<br>
<br>
After an exception has been handled the program, execution resumes after the <i>try-catch</i> block, not after the <code>throw</code> statement!.<br>
<br>
It is also possible to nest <code>try-catch</code> blocks within more external <code>try</code> blocks. In these cases, we have the possibility that an internal <code>catch</code> block forwards the exception to its external level. This is done with the expression <code>throw;</code> with no arguments. For example: <br>
<br>
<div class="auto"><table class="snippet"><tr><td class="rownum"><pre><code>1<br>2<br>3<br>4<br>5<br>6<br>7<br>8<br>9<br>10<br>11<br></code></pre></td>
<td class="source"><pre><code><var>try</var> {
  <var>try</var> {
      <cite>// code here</cite>
  }
  <var>catch</var> (<var>int</var> n) {
      <var>throw</var>;
  }
}
<var>catch</var> (...) {
  cout &lt;&lt; <kbd>"Exception occurred"</kbd>;
}</code></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
</section>
<br>
<section id="exception_specification">
<h3>Exception specification</h3>
Older code may contain <i>dynamic exception specifications</i>. They are now deprecated in C++, but still supported. A <i>dynamic exception specification</i> follows the declaration of a function, appending a <code>throw</code> specifier to it. For example:<br>
<br>
<div class="auto"><table class="snippet"><tr><td class="rownum"><pre><code>&nbsp;</code></pre></td>
<td class="source"><pre><code><var>double</var> myfunction (<var>char</var> param) <var>throw</var> (<var>int</var>);</code></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
<br>
This declares a function called <code>myfunction</code>, which takes one argument of type <code>char</code> and returns a value of type <code>double</code>. If this function throws an exception of some type other than <code>int</code>, the function calls <samp><a href="/unexpected">std::unexpected</a></samp> instead of looking for a handler or calling <samp><a href="/terminate">std::terminate</a></samp>.<br>
<br>
If this <code>throw</code> specifier is left empty with no type, this means that <samp><a href="/unexpected">std::unexpected</a></samp> is called for any exception. Functions with no <code>throw</code> specifier (regular functions) never call <samp><a href="/unexpected">std::unexpected</a></samp>, but follow the normal path of looking for their exception handler.<br>
<br>
<div class="auto"><table class="snippet"><tr><td class="rownum"><pre><code>1<br>2<br></code></pre></td>
<td class="source"><pre><code><var>int</var> myfunction (<var>int</var> param) <var>throw</var>(); <cite>// all exceptions call unexpected</cite>
<var>int</var> myfunction (<var>int</var> param);         <cite>// normal exception handling </cite></code></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
</section>
<br>
<section id="standard_exceptions">
<h3>Standard exceptions</h3>
The C++ Standard library provides a base class specifically designed to declare objects to be thrown as exceptions. It is called <code><a href="/exception">std::exception</a></code> and is defined in the <code><a href="/&lt;exception&gt;">&lt;exception&gt;</a></code> header. This class has a virtual member function called <code>what</code> that returns a null-terminated character sequence (of type <code>char *</code>) and that can be overwritten in derived classes to contain some sort of description of the exception.<br>
<br>
<div class="auto"><table class="split"><tr><td class="rownum"><pre><code>1<br>2<br>3<br>4<br>5<br>6<br>7<br>8<br>9<br>10<br>11<br>12<br>13<br>14<br>15<br>16<br>17<br>18<br>19<br>20<br>21<br>22<br>23<br>24<br></code></pre></td>
<td class="source"><pre><code><cite>// using standard exceptions</cite>
<dfn>#include &lt;iostream&gt;</dfn>
<dfn>#include &lt;exception&gt;</dfn>
<var>using</var> <var>namespace</var> std;

<var>class</var> myexception: <var>public</var> exception
{
  <var>virtual</var> <var>const</var> <var>char</var>* what() <var>const</var> <var>throw</var>()
  {
    <var>return</var> <kbd>"My exception happened"</kbd>;
  }
} myex;

<var>int</var> main () {
  <var>try</var>
  {
    <var>throw</var> myex;
  }
  <var>catch</var> (exception&amp; e)
  {
    cout &lt;&lt; e.what() &lt;&lt; <kbd>'\n'</kbd>;
  }
  <var>return</var> 0;
}</code></pre></td>
<td class="output"><pre><samp>My exception happened.</samp></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
<br>
We have placed a handler that catches exception objects by reference (notice the ampersand <code>&</code> after the type), therefore this catches also classes derived from <code>exception</code>, like our <code>myex</code> object of type <code>myexception</code>.<br>
<br>
All exceptions thrown by components of the C++ Standard library throw exceptions derived from this <code>exception</code> class. These are:<br>
<br>
<table class="boxed">
<tr><th>exception</th><th>description</th></tr>
<tr><td><code><a href="/bad_alloc">bad_alloc</a></code></td><td>thrown by <code>new</code> on allocation failure</td></tr>
<tr><td><code><a href="/bad_cast">bad_cast</a></code></td><td>thrown by <code>dynamic_cast</code> when it fails in a dynamic cast</td></tr>
<tr><td><code><a href="/bad_exception">bad_exception</a></code></td><td>thrown by certain dynamic exception specifiers</td></tr>
<tr><td><code><a href="/bad_typeid">bad_typeid</a></code></td><td>thrown by <code>typeid</code></td></tr>
<tr><td><code><a href="/bad_function_call">bad_function_call</a></code></td><td>thrown by empty <code><a href="/function">function</a></code> objects</td></tr>
<tr><td><code><a href="/bad_weak_ptr">bad_weak_ptr</a></code></td><td>thrown by <code><a href="/shared_ptr">shared_ptr</a></code> when passed a bad <code><a href="/weak_ptr">weak_ptr</a></code></td></tr>
</table>
<br>
Also deriving from <code>exception</code>, header <code><a href="/&lt;exception&gt;">&lt;exception&gt;</a></code> defines two generic exception types that can be inherited by custom exceptions to report errors:<br>
<br>
<table class="boxed">
<tr><th>exception</th><th>description</th></tr>
<tr><td><code><a href="/logic_error">logic_error</a></code></td><td>error related to the internal logic of the program</td></tr>
<tr><td><code><a href="/runtime_error">runtime_error</a></code></td><td>error detected during runtime</td></tr>
</table>
<br>
A typical example where standard exceptions need to be checked for is on memory allocation:<br>
<br>
<div class="auto"><table class="split"><tr><td class="rownum"><pre><code>1<br>2<br>3<br>4<br>5<br>6<br>7<br>8<br>9<br>10<br>11<br>12<br>13<br>14<br>15<br>16<br></code></pre></td>
<td class="source"><pre><code><cite>// bad_alloc standard exception</cite>
<dfn>#include &lt;iostream&gt;</dfn>
<dfn>#include &lt;exception&gt;</dfn>
<var>using</var> <var>namespace</var> std;

<var>int</var> main () {
  <var>try</var>
  {
    <var>int</var>* myarray= <var>new</var> <var>int</var>[1000];
  }
  <var>catch</var> (exception&amp; e)
  {
    cout &lt;&lt; <kbd>"Standard exception: "</kbd> &lt;&lt; e.what() &lt;&lt; endl;
  }
  <var>return</var> 0;
}</code></pre></td>
<td class="output"><pre><samp> </samp></pre></td><td class="C_btnholder"></td></tr></table></div>
<br>
<br>
The exception that may be caught by the exception handler in this example is a <code>bad_alloc</code>. Because <code>bad_alloc</code> is derived from the standard base class <code>exception</code>, it can be caught (capturing by reference, captures all related classes).<br>
</section><table class="C_docPrevNext"><tr><td class="prev"><a href="/doc/tutorial/typecasting/" rel="prev"><img src="/ico/32arrowleft.png" width="32" height="32" alt="Previous">Previous:<br><b>Type conversions</b></a></td><td class="index"><a href="/doc/tutorial/" rel="index"><img src="/ico/32doc.png" width="32" height="32" alt="Index"><br>Index</a></td><td class="next"><a href="/doc/tutorial/preprocessor/" rel="next"><img src="/ico/32arrowright.png" width="32" height="32" alt="Next">Next:<br><b>Preprocessor directives</b></a></td></tr></table></div><div id="CH_bb"></div></div>
<script type="text/javascript">
    google_ad_client = "ca-pub-1444228343479937";
    google_ad_slot = "9701143201";
    google_ad_width = 728;
    google_ad_height = 90;
</script>
<!-- leaderboard -->
<script type="text/javascript"
src="//pagead2.googlesyndication.com/pagead/show_ads.js">
</script></div>
<div id="I_nav">
<div class="sect root">
<h3><b><a href="/">C++</a></b></h3>
<ul>
<li class="folder info"><a href="/info/">Information</a></li>
<li class="folder selected doc"><a href="/doc/">Tutorials</a></li>
<li class="folder reference"><a href="/reference/">Reference</a></li>
<li class="folder articles"><a href="/articles/">Articles</a></li>
<li class="folder forum"><a href="/forum/">Forum</a></li>
</ul>
</div>
<div class="C_BoxLabels C_BoxSort sect" id="doc_box">
<h3><b><a href="/doc/">Tutorials</a></b></h3>
<ul>
<li class="cpp98 cpp11 selected folder"><a href="/doc/tutorial/"><span>C++ Language</span></a></li>
<li><a href="/doc/ascii/"><span>Ascii Codes</span></a></li>
<li><a href="/doc/boolean/"><span>Boolean Operations</span></a></li>
<li><a href="/doc/hex/"><span>Numerical Bases</span></a></li>
</ul>
</div>
<div class="C_BoxLabels C_BoxSort sect cpp98 cpp11" id="tutorial_box">
<h3><b><a href="/doc/tutorial/">C++ Language</a></b></h3>
<ul>
<li class="C_Collapsable"><h4>Introduction:</h4><ul>
<li class="folder"><a href="/doc/tutorial/introduction/"><span>Compilers</span></a></li>
</ul></li>
<li class="C_Collapsable"><h4>Basics of C++:</h4><ul>
<li><a href="/doc/tutorial/program_structure/"><span>Structure of a program</span></a></li>
<li><a href="/doc/tutorial/variables/"><span>Variables and types</span></a></li>
<li><a href="/doc/tutorial/constants/"><span>Constants</span></a></li>
<li><a href="/doc/tutorial/operators/"><span>Operators</span></a></li>
<li><a href="/doc/tutorial/basic_io/"><span>Basic Input/Output</span></a></li>
</ul></li>
<li class="C_Collapsable"><h4>Program structure:</h4><ul>
<li><a href="/doc/tutorial/control/"><span>Statements and flow control</span></a></li>
<li><a href="/doc/tutorial/functions/"><span>Functions</span></a></li>
<li><a href="/doc/tutorial/functions2/"><span>Overloads and templates</span></a></li>
<li><a href="/doc/tutorial/namespaces/"><span>Name visibility</span></a></li>
</ul></li>
<li class="C_Collapsable"><h4>Compound data types:</h4><ul>
<li><a href="/doc/tutorial/arrays/"><span>Arrays</span></a></li>
<li><a href="/doc/tutorial/ntcs/"><span>Character sequences</span></a></li>
<li><a href="/doc/tutorial/pointers/"><span>Pointers</span></a></li>
<li><a href="/doc/tutorial/dynamic/"><span>Dynamic memory</span></a></li>
<li><a href="/doc/tutorial/structures/"><span>Data structures</span></a></li>
<li><a href="/doc/tutorial/other_data_types/"><span>Other data types</span></a></li>
</ul></li>
<li class="C_Collapsable"><h4>Classes:</h4><ul>
<li><a href="/doc/tutorial/classes/"><span>Classes (I)</span></a></li>
<li><a href="/doc/tutorial/templates/"><span>Classes (II)</span></a></li>
<li><a href="/doc/tutorial/classes2/"><span>Special members</span></a></li>
<li><a href="/doc/tutorial/inheritance/"><span>Friendship and inheritance</span></a></li>
<li><a href="/doc/tutorial/polymorphism/"><span>Polymorphism</span></a></li>
</ul></li>
<li class="C_Collapsable"><h4>Other language features:</h4><ul>
<li><a href="/doc/tutorial/typecasting/"><span>Type conversions</span></a></li>
<li class="selected"><a href="/doc/tutorial/exceptions/"><span>Exceptions</span></a></li>
<li><a href="/doc/tutorial/preprocessor/"><span>Preprocessor directives</span></a></li>
</ul></li>
<li class="C_Collapsable"><h4>Standard library:</h4><ul>
<li><a href="/doc/tutorial/files/"><span>Input/output with files</span></a></li>
</ul></li>
</ul>
</div>
<div id="I_subnav"></div>
<div class="C_ad234">
<script type="text/javascript"><!--
google_ad_client = "ca-pub-7688470879129516";
google_ad_slot = "7445514729";
google_ad_width = 234;
google_ad_height = 60;
//-->
</script>
<script type="text/javascript"
src="http://pagead2.googlesyndication.com/pagead/show_ads.js">
</script>
</div></div>
<div id="I_midclear"></div>
</div>
</div>
<div id="I_bottom">
<div id="I_footer">
	<a href="/">Home page</a> | <a href="/privacy.do">Privacy policy</a><br>&copy; cplusplus.com, 2000-2016 - All rights reserved - <i>v3.1</i><br><a href="/contact.do?referrer=www.cplusplus.com%2Fdoc%2Ftutorial%2Fexceptions%2F">Spotted an error? contact us</a>
</div>
</div>

<script type="text/javascript">
<!--
 onSession ( function(us) {
		el=document.getElementById('CH_bb'); el.innerHTML='';
		if (us.auth(4096)) {
			
			el.appendChild(btn('Edit Page','Edit Page','edit','/doc/edit.cgi?a=tutorial+exceptions'));
			el.appendChild(btn('Delete Page','Delete Page','delete','javascript:docdel()'));
		}
	});
	function docdel() { if (confirm('WARNING: You are about to delete this page. Confirm?')) window.location='/doc/delete.cgi?a=tutorial+exceptions'; }
	ready();
var _gaq = _gaq || [];
_gaq.push(['_setAccount', 'UA-521783-1']);
_gaq.push(['_trackPageview']);

(function() {
  var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
  ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
  var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
})();

//-->
</script>

</body>
</html>